Spatnav rework

frivoal opened the pull request:
This pull request makes a number of substantial modification to the specification, based on discussions the various discussions we've had since completing the first draft.

This closes or addresse a quite a few open issues, as detailed in the individual commit message.

The overall effect of these changes is to simplify the specification, and make it more focused on its core purpose. Editorial changes should also better highlight the essential parts.

Most of these changes are conceptually independent, but the corresponding edits overlap. Since I believe we're in general agreement over all of these, making separate pull requests for each would be quite a pain for little gain, so I made a single pull request.

----

Note that index.html needs to be regenerated after merging.

hugoholgersson commented on index.bs:
```diff
@@ -640,20 +644,26 @@ dictionary SelectSpatNavBestCandidateOptions {
 
 partial interface Element {
-    sequence&lt;Node> findSpatNavCandidates(FindSpatNavCandidatesOptions arg);
+    Node getSpatnavContainer();
```
I understand this is only a convenience method... but.... getting the "spatnav container" can be done with vanilla JS...

Isn't it better to leave polyfillable things out from the spec? At least until JS framework ask for it being done natively.

In principle, I'd like the spec to focus on adding/enabling new use case that's not already possible in plain JS... :) 

If you do have to add this method, can't it be called something that hints for use cases that aren't related to spatnav. Maybe "enclosingScoller()"?

hugoholgersson replied:
In Blink we call it "scrollableAreaOrDocumentOf()"... How does that sound?

frivoal replied:
As you say, it can be done in JS. But it is a nice convenience. How about meeting in the middle, and keeping it for now, but marking it "at risk" to signal to implementors that we don't consider this a MUST HAVE, and are ok with dropping it if they don't want to implement.

As for the name, I'd rather not use `enclosingScroller()` or `scrollableAreaOrDocumentOf()`, because these names would become misleading when the `spatnav-container` property, or any other mechanism to turn non scrollers into spatnav containers.

hugoholgersson replied:
I thought we were removing the `spatnav-container` property.

OK for marking it "at risk".

hugoholgersson commented on index.bs:
```diff
@@ -640,20 +644,26 @@ dictionary SelectSpatNavBestCandidateOptions {
 partial interface Element {
-    sequence&lt;Node> findSpatNavCandidates(FindSpatNavCandidatesOptions arg);
+    Node getSpatnavContainer();
+    sequence&lt;Node> focusableAreas(optional FocusableAreasOptions arg);
```
Like!

hugoholgersson commented on index.bs:
```diff
@@ -638,16 +638,17 @@ dictionary FocusableAreasOptions {
     sequence&lt;Node> focusableAreas(optional FocusableAreasOptions arg);
-    Node? selectSpatNavBestCandidate(SelectSpatNavBestCandidateOptions arg);
-    Node? sequentialNavSearch(SequentialNavigationDirection dir);
+    Node? spatNavSearch(SpatNavSearchOptions arg);
```
As said in #37, I really think these two methods should be merged to one method.

In Blink we've got the WebFocusType-enum that contains both sequential and spatial navigation "directions". Why not here?

hugoholgersson replied:
Sure!

frivoal commented on index.bs:
```diff
@@ -119,18 +120,18 @@ Overview</h2>
-(by default the root element, scrollable elements, and iframes,
-but other elements can be made into <a>spatial navigation focus containers</a>
-using the 'spatnav-container' property).
+(the root element, scrollable elements, or iframes)
```
I agree we should have that property eventually, and the behavior (assuming we don't change the processsing model) is straightforward. So I am not against having it in level 1.

At the same time, we can easily add it back at any time given that it's compatible with the overall model. That makes it non-essential, and since a focused spec is good for making sure people get the point, and to limit implementation scope, I thought dropping it for now made sense.

If you'd like to keep it, I agree with putting it somewhere further down the spec, maybe after the navigation part, or even after the processing model. And in that case, yes, this sentence would need to be reverted. Should I do so, or should we keep it for level 2?

jihyerish replied:
For now, I'd like to keep it in level 1.

I think the essential part of the level 1 is the Processing Model as well as your opinion.
Processing Model uses "container" in general, and it is the vital concept of it.
There are ways to allow author-defined focus navigation heuristic and handling keys via JS API and NavigationEvent.
It also seems nice to have the author-defined container by spatnav-container.

Therefore, can we revert this sentence?

frivoal replied:
It is possible for the author to define containers by themselves without this property, using javascript: catch the events, decide on what they want to use as the container, use `focusableAreas()` on that element, and ` spatNavSearch()` with the resulting list of candidates. So no capability is lost by dropping this property.

But it is indeed much simpler to do this with a CSS property, So I'm OK with putting it back (and revert the sentence). I would like to mark it "at risk" though, like other things that "for now we would like to keep in level 1" (as opposed to things that must absolutely be in level 1 or the spec completely changes).

hugoholgersson replied:
I agree with Florian here. I think we should try to avoid adding too many "convenience-helpers" for things that can be done in other ways. I'd Level 1 to only contain the smallest set of primitives needed (à l'Extensible Web).

hugoholgersson replied:
( In general, isn't the `navigate`-[event](https://github.com/WICG/spatial-navigation/issues/32#issuecomment-372963209) the only primitive we need to polyfill all other functionally that's been suggested in this repo? If so, I'd say implementing `navigate` should be this spec's "Level 0". )

jihyerish replied:
I realized that thoughts about core feature(the set of primitives) of SpatNav Level1 are different.
I'm with you about 'spatnav-container' can be implemented with JS.

But, while I implement the polyfill of SpatNav, I'm certain about having the customized container is convenient. Without it, all trial of spatial navigation needs to bring all candidates of the document.
Also if this is implemented with JS, it will become a burden for the performance.

Therefore I think 'spatnav-container" is essential "convenience-helper" for SpatNav Level 1.

jihyerish replied:
@hugoholgersson Could you share your thought about the primitives of Level 1?

hugoholgersson replied:
> Also if this is implemented with JS, it will become a burden for the performance.

In JS, only one DOM-height climb needs to be done. And that's very cheap. Blink does it all the time already, for example on every mouse movement.

```
 // Polyfilled spatnav-container: 
function handleNavigation(event) {
  if (event.target == mySpatnavContainer && !mySpatnavContainer.contains(e.relatedTarget))
    preventDefault(); // Jail navigation.
}
```

> Without it, all the trials of spatial navigation need to bring all candidates of the document.

You're right...  Every key stroke's we will check all of the document's focusables (in worst case only). I'm not sure that performance difference is big enough to motivate a new CSS property and all the logic that comes with it.

> Could you share your thought about the primitives of Level 1?

I think a navigate-event is the only primitive authors would need to experiment and come up with better constructs and smarter convenience-helpers than we can. I'd leave everything else out :)

frivoal replied:
I think that lots of people (myself included) see a level 0 hiding inside this level 1 as a subset of what's in there. The problem is that we all seem to see a different subset.

For now, I'm dealing with this with marking possibly non essential parts "at-risk"

frivoal commented:
Alright, the latest update to this PR addresses the feedback raised in the comments. It may not be perfect yet, but it is an improvement over what was there before. I'm merging it, and we can open individual issues for anything that's still a problem.
